Firestore/core/src/core/composite_filter.cc

```diff
@@ -0,0 +1,162 @@
+      return "and";
+    default:
+      UNREACHABLE();
+  }
+}
+
+}  // namespace
+
+CompositeFilter CompositeFilter::Create(const std::vector<Filter>&& filters,
+                                        Operator op) {
+  return CompositeFilter(
+      std::make_shared<const Rep>(Rep(std::move(filters), op)));
+}
+
+CompositeFilter::CompositeFilter(std::shared_ptr<const Filter::Rep> rep)
+    : Filter(std::move(rep)) {
+}
+
+CompositeFilter::Rep::Rep(const std::vector<Filter>&& filters, Operator op)
+    : filters_(std::move(filters)), op_(op) {
+}
+
+CompositeFilter::CompositeFilter(const Filter& other) : Filter(other) {
+  HARD_ASSERT(other.IsACompositeFilter());
+}
+
+bool CompositeFilter::Rep::IsConjunction() const {
+  return op_ ==
+         Operator::
+             google_firestore_v1_StructuredQuery_CompositeFilter_Operator_AND;
+}
+
+bool CompositeFilter::Rep::IsDisjunction() const {
+  // TODO(orquery): Replace with Operator.OR.
+  return op_ == Operator::google_firestore_v1_StructuredQuery_\
+CompositeFilter_Operator_OPERATOR_UNSPECIFIED;
+}
+
+bool CompositeFilter::Rep::Matches(const model::Document& doc) const {
+  if (IsConjunction()) {
+    // For conjunctions, all filters must match, so return false if any filter
+    // doesn't match.
+    for (const auto& filter : filters_) {
+      if (!filter.Matches(doc)) {
+        return false;
+      }
+    }
+    return true;
+  } else {
+    // For disjunctions, at least one filter should match.
+    for (const auto& filter : filters_) {
+      if (filter.Matches(doc)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+std::string CompositeFilter::Rep::CanonicalId() const {
+  // TODO(orquery): Add special case for flat AND filters.
+  return util::StringFormat(
+      "%s(%s)", CanonicalName(op_),
+      absl::StrJoin(filters_, ",", [](std::string* out, const Filter& f) {
+        return absl::StrAppend(out, f.CanonicalId());
+      }));
+}
+
+bool CompositeFilter::Rep::Equals(const Filter::Rep& other) const {
+  if (!other.IsACompositeFilter()) return false;
+  const auto& other_rep = static_cast<const CompositeFilter::Rep&>(other);
+  // Note: This comparison requires order of filters in the list to be the same,
+  // and it does not remove duplicate subfilters from each composite filter.
+  // It is therefore way less expensive.
+  // TODO(orquery): Consider removing duplicates and ignoring order of filters
+  // in the list.
+  return op_ == other_rep.op_ && filters_ == other_rep.filters_;
+}
+
+const FieldFilter* CompositeFilter::Rep::FindFirstMatchingFilter(
+    CheckingFun& condition) const {
+  for (const auto& field_filter : GetFlattenedFilters()) {
+    if (condition(field_filter)) {
+      return &field_filter;
+    }
+  }
+  return nullptr;
+}
+
+const model::FieldPath* CompositeFilter::Rep::GetFirstInequalityField() const {
+  CheckingFun condition = [](const FieldFilter& field_filter_ptr) {
```

Comment:
it's not really a pointer. so `field_filter` is probably a better variable name.